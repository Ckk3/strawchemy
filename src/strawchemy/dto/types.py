"""DTO domain types."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping


__all__ = ("DTOConfig", "DTOFieldConfig", "Purpose")


class DTOMissingType:
    """A sentinel type to detect if a parameter is supplied or not when.

    constructing pydantic FieldInfo.
    """


DTO_MISSING = DTOMissingType()


class Purpose(str, Enum):
    """For identifying the purpose of a DTO to the factory.

    The factory will exclude fields marked as private or read-only on the domain model depending
    on the purpose of the DTO.

    Example:
    ```python
    ReadDTO = dto.factory("AuthorReadDTO", Author, purpose=dto.Purpose.READ)
    ```
    """

    READ = "read"
    """To mark a DTO that is to be used to serialize data returned to
    clients."""
    WRITE = "write"
    """To mark a DTO that is to deserialize and validate data provided by
    clients."""
    COMPLETE = "complete"
    """To mark a DTO that is to deserialize and validate data provided by
    clients. Fields marked as TO_COMPLETE must not be null."""


@dataclass
class PurposeConfig:
    """Mark the field as read-only, or private."""

    type_override: Any | None = DTO_MISSING
    validator: Callable[[Any], Any] | None = None
    """Single argument callables that are defined on the DTO as validators for the field."""
    alias: str | None = None
    """Customize name of generated pydantic field."""
    partial: bool | None = None


@dataclass
class DTOFieldConfig:
    """For configuring DTO behavior on SQLAlchemy model fields."""

    purposes: set[Purpose] = field(default_factory=lambda: {Purpose.READ, Purpose.WRITE})
    default_config: PurposeConfig = field(default_factory=PurposeConfig)
    configs: dict[Purpose, PurposeConfig] = field(default_factory=dict)

    def purpose_config(self, dto_config: DTOConfig) -> PurposeConfig:
        return self.configs.get(dto_config.purpose, self.default_config)


@dataclass
class DTOConfig:
    """Control the generated DTO."""

    purpose: Purpose
    """Configure the DTO for "read" or "write" operations."""
    exclude: set[str] = field(default_factory=set)
    """Explicitly exclude fields from the generated DTO."""
    partial: bool | None = None
    """Make all field optional."""
    type_map: Mapping[Any, Any] = field(default_factory=dict)

    aliases: Mapping[str, str] = field(default_factory=dict)

    alias_generator: Callable[[str], str] | None = None

    def __post_init__(self) -> None:
        if self.aliases and self.alias_generator is not None:
            msg = "You must set `aliases` or `alias_generator`, not both"
            raise ValueError(msg)

    def alias(self, name: str) -> str | None:
        if self.aliases:
            return self.aliases.get(name)
        if self.alias_generator is not None:
            return self.alias_generator(name)
        return None

    @property
    def cache_key(self) -> int:
        return hash(
            (self.purpose, self.partial, self.alias_generator, tuple(self.type_map), tuple(self.type_map.values()))
        )
