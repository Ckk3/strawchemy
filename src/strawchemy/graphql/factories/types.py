from __future__ import annotations

import dataclasses
from collections.abc import Generator
from typing import TYPE_CHECKING, Any, TypeVar, override

from sqlalchemy.orm import DeclarativeBase, QueryableAttribute
from strawchemy.dto.backend.strawberry import StrawberrryDTOBackend
from strawchemy.dto.base import (
    DTOBackend,
    DTOBase,
    DTOFactory,
    DTOFieldDefinition,
    Relation,
)
from strawchemy.dto.types import DTO_MISSING, DTOConfig, Purpose
from strawchemy.graph import Node
from strawchemy.graphql import typing as strawchemy_typing
from strawchemy.graphql.constants import AGGREGATIONS_KEY, NODES_KEY
from strawchemy.graphql.dto import (
    AggregateDTO,
    AggregateFieldDefinition,
    DTOKey,
    EnumDTO,
    FunctionFieldDefinition,
    GraphQLFieldDefinition,
)
from strawchemy.graphql.typing import AggregateDTOT, GraphQLDTOT

from .aggregations import AggregationInspector
from .enum import EnumDTOFactory

if TYPE_CHECKING:
    from collections.abc import Generator

    from strawchemy.graph import Node
    from strawchemy.graphql.inspector import GraphQLInspectorProtocol
    from strawchemy.sqlalchemy.typing import DeclarativeT


__all__ = ("AggregateDTOFactory", "DistinctOnFieldsDTOFactory", "RootAggregateTypeDTOFactory", "TypeDTOFactory")

T = TypeVar("T")

_TYPING_NS = vars(strawchemy_typing)


class GraphQLDTOFactory(DTOFactory[DeclarativeBase, QueryableAttribute[Any], GraphQLDTOT]):
    inspector: GraphQLInspectorProtocol

    def type_description(self) -> str:
        return "GraphQL type"

    @override
    def type_hint_namespace(self) -> dict[str, Any]:
        return super().type_hint_namespace() | _TYPING_NS

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(name, model, dto_config, base, node, raise_if_no_fields, **kwargs):
            key = DTOKey.from_dto_node(node)
            graphql_field = GraphQLFieldDefinition.from_field(field)
            yield graphql_field
            field_map[key + field.name] = graphql_field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        field_map: dict[DTOKey, GraphQLFieldDefinition] = {}
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            field_map=field_map,
            **kwargs,
        )
        if not dto.__strawchemy_field_map__:
            dto.__strawchemy_field_map__ = field_map
        dto.__strawchemy_description__ = self.type_description()
        return dto


class TypeDTOFactory(GraphQLDTOFactory[GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_factory: AggregateDTOFactory[AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, StrawberrryDTOBackend(AggregateDTO)
        )

    def _aggregation_field(
        self, field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], dto_config: DTOConfig
    ) -> GraphQLFieldDefinition:
        related_model = self.inspector.relation_model(field_def.model_field)
        aggregate_dto_config = dataclasses.replace(dto_config, annotation_overrides={})
        dto = self._aggregation_factory.factory(
            model=related_model, dto_config=aggregate_dto_config, parent_field_def=field_def
        )
        return AggregateFieldDefinition(
            dto_config=dto_config,
            model=dto.__dto_model__,
            _model_field=field_def.model_field,
            model_field_name=f"{field_def.name}_aggregate",
            type_hint=dto,
            related_dto=dto,
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}{'Input' if dto_config.purpose is Purpose.WRITE else ''}Type"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(
            name, model, dto_config, base, node, raise_if_no_fields, field_map=field_map, **kwargs
        ):
            key = DTOKey.from_dto_node(node)
            if field.is_relation and field.uselist and aggregations:
                aggregation_field = self._aggregation_field(field, dto_config)
                field_map[key + aggregation_field.name] = aggregation_field
                yield aggregation_field
            yield field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        return super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations if dto_config.purpose is Purpose.READ else False,
            **kwargs,
        )


class RootAggregateTypeDTOFactory(TypeDTOFactory[GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        type_factory: TypeDTOFactory[GraphQLDTOT] | None = None,
        aggregation_factory: AggregateDTOFactory[AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._type_factory = type_factory or TypeDTOFactory(inspector, backend)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, StrawberrryDTOBackend(AggregateDTO)
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Root"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[DeclarativeT],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        if not node.is_root:
            yield from ()
        key = DTOKey.from_dto_node(node)
        field_map = field_map if field_map is not None else {}
        nodes_dto = self._type_factory.factory(model, dto_config=dto_config, aggregations=aggregations)
        nodes = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=NODES_KEY,
            type_hint=list[nodes_dto],
            is_relation=False,
        )
        aggregations_field = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=AGGREGATIONS_KEY,
            type_hint=self._aggregation_factory.factory(model, dto_config=dto_config),
            is_relation=False,
            is_aggregate=True,
        )
        field_map[key + nodes.name] = nodes
        field_map[key + aggregations_field.name] = aggregations_field
        yield from iter((nodes, aggregations_field))

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations,
            **kwargs,
        )
        dto.__strawchemy_is_root_aggregation_type__ = True
        return dto


class AggregateDTOFactory(GraphQLDTOFactory[AggregateDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[AggregateDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_builder: AggregationInspector | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_builder = aggregation_builder or AggregationInspector(inspector)

    @override
    def type_description(self) -> str:
        return "Aggregation fields"

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, AggregateDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Aggregate"

    @override
    def _factory(
        self,
        name: str,
        model: type[DeclarativeT],
        dto_config: DTOConfig,
        node: Node[Relation[Any, AggregateDTOT], None],
        base: type[Any] | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> type[AggregateDTOT]:
        field_map = field_map if field_map is not None else {}
        model_field = parent_field_def.model_field if parent_field_def else None
        as_partial_config = dataclasses.replace(dto_config, partial=True)
        field_definitions: list[FunctionFieldDefinition] = [
            FunctionFieldDefinition(
                dto_config=dto_config,
                model=model,
                _model_field=model_field if model_field is not None else DTO_MISSING,
                model_field_name=aggregation.function,
                type_hint=aggregation.output_type,
                _function=aggregation,
                default=aggregation.default,
            )
            for aggregation in self._aggregation_builder.output_functions(model, as_partial_config)
        ]

        root_key = DTOKey.from_dto_node(node)
        field_map.update({root_key + field.model_field_name: field for field in field_definitions})
        return self.backend.build(name, model, field_definitions, **(backend_kwargs or {}))


class DistinctOnFieldsDTOFactory(EnumDTOFactory):
    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, EnumDTO], None] | None = None
    ) -> str:
        return f"{base_name}DistinctOnFields"
