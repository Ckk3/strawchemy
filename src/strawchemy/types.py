from __future__ import annotations

from dataclasses import dataclass

__all__ = ("DefaultOffsetPagination",)


@dataclass(eq=True, frozen=True)
class DefaultOffsetPagination:
    limit: int | None = None
    offset: int = 0
