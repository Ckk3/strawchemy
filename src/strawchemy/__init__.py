"""Custom DTO implementation."""

from __future__ import annotations

from .mapper import Strawchemy
from .sqlalchemy.hook import FilterOrderHook, LoadColumnsHook, QueryHookProtocol
from .strawberry import ModelInstance
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository

__all__ = (
    "FilterOrderHook",
    "LoadColumnsHook",
    "ModelInstance",
    "QueryHookProtocol",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemySyncRepository",
)
