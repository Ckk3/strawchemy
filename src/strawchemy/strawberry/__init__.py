from __future__ import annotations

from strawchemy.types import DefaultOffsetPagination

from ._field import StrawchemyCreateUpdateMutationField, StrawchemyDeleteMutationField, StrawchemyField
from ._instance import ModelInstance
from ._utils import default_session_getter

__all__ = (
    "DefaultOffsetPagination",
    "ModelInstance",
    "StrawchemyCreateUpdateMutationField",
    "StrawchemyDeleteMutationField",
    "StrawchemyField",
    "default_session_getter",
)
