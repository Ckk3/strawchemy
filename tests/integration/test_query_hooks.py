from __future__ import annotations

from typing import TYPE_CHECKING

import pytest
from strawchemy import StrawchemyAsyncRepository, StrawchemySyncRepository

import strawberry
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .types import (
    ColorTypeHooks,
    ColorWithFilteredFruit,
    FilteredFruitType,
    FruitTypeHooks,
    OrderedFruitType,
    strawchemy,
)
from .typing import RawRecordData

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion

    from .fixtures import QueryTracker

pytestmark = [pytest.mark.integration]


@strawberry.type
class AsyncQuery:
    fruits: list[FruitTypeHooks] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    fruits_paginated: list[FruitTypeHooks] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )
    filtered_fruits: list[FilteredFruitType] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    filtered_fruits_paginated: list[FilteredFruitType] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )
    ordered_fruits: list[OrderedFruitType] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    ordered_fruits_paginated: list[OrderedFruitType] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )
    colors: list[ColorWithFilteredFruit] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    colors_paginated: list[ColorWithFilteredFruit] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )
    colors_hooks: list[ColorTypeHooks] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    colors_hooks_paginated: list[ColorTypeHooks] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )


@strawberry.type
class SyncQuery:
    fruits: list[FruitTypeHooks] = strawchemy.field(repository_type=StrawchemySyncRepository)
    fruits_paginated: list[FruitTypeHooks] = strawchemy.field(repository_type=StrawchemySyncRepository, pagination=True)
    filtered_fruits: list[FilteredFruitType] = strawchemy.field(repository_type=StrawchemySyncRepository)
    filtered_fruits_paginated: list[FilteredFruitType] = strawchemy.field(
        repository_type=StrawchemySyncRepository, pagination=True
    )
    ordered_fruits: list[OrderedFruitType] = strawchemy.field(repository_type=StrawchemySyncRepository)
    ordered_fruits_paginated: list[OrderedFruitType] = strawchemy.field(
        repository_type=StrawchemySyncRepository, pagination=True
    )
    colors: list[ColorWithFilteredFruit] = strawchemy.field(repository_type=StrawchemySyncRepository)
    colors_paginated: list[ColorWithFilteredFruit] = strawchemy.field(
        repository_type=StrawchemySyncRepository, pagination=True
    )

    colors_hooks: list[ColorTypeHooks] = strawchemy.field(repository_type=StrawchemySyncRepository)
    colors_hooks_paginated: list[ColorTypeHooks] = strawchemy.field(
        repository_type=StrawchemySyncRepository, pagination=True
    )


@pytest.fixture
def sync_query() -> type[SyncQuery]:
    return SyncQuery


@pytest.fixture
def async_query() -> type[AsyncQuery]:
    return AsyncQuery


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
@pytest.mark.snapshot
async def test_load_columns_hook(
    fruits_query: str,
    any_query: AnyQueryExecutor,
    raw_fruits: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ description }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[fruits_query] == [
        {"description": f"The {fruit['name']} color id is {fruit['color_id']}"} for fruit in raw_fruits
    ]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
@pytest.mark.snapshot
async def test_load_relationships_with_columns(
    fruits_query: str,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    raw_fruits: RawRecordData,
    raw_colors: RawRecordData,
) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ prettyColor }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[fruits_query] == [
        {"prettyColor": f"Color is {next(color['name'] for color in raw_colors if color['id'] == fruit['color_id'])}"}
        for fruit in raw_fruits
    ]

    query_tracker.assert_statements(2, "select", sql_snapshot)


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
@pytest.mark.snapshot
async def test_load_relationships_no_columns(
    fruits_query: str,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    raw_fruits: RawRecordData,
) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ prettyFarms }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[fruits_query] == [{"prettyFarms": f"Farms are: {fruit['name']} farm"} for fruit in raw_fruits]

    query_tracker.assert_statements(2, "select", sql_snapshot)


@pytest.mark.parametrize("query", ["colors", "colorsPaginated"])
@pytest.mark.snapshot
async def test_load_relationships_nested(
    query: str,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    raw_colors: RawRecordData,
    raw_fruits: RawRecordData,
) -> None:
    result = await maybe_async(any_query(f"{{ {query} {{ farms }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[query] == [
        {
            "farms": f"Farms are: {', '.join(f'{fruit["name"]} farm' for fruit in raw_fruits if fruit['color_id'] == color['id'])}"
        }
        for color in raw_colors
    ]

    query_tracker.assert_statements(2, "select", sql_snapshot)


@pytest.mark.parametrize("query", ["colorsHooks", "colorsHooksPaginated"])
@pytest.mark.snapshot
async def test_load_relationships_on_nested_field(
    query: str,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    raw_colors: RawRecordData,
    raw_fruits: RawRecordData,
) -> None:
    result = await maybe_async(any_query(f"{{ {query} {{ id fruits {{ prettyColor }} }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[query] == [
        {
            "id": color["id"],
            "fruits": [
                {"prettyColor": f"Color is {color['name']}"}
                for _ in range(len([fruit for fruit in raw_fruits if fruit["color_id"] == color["id"]]))
            ],
        }
        for color in raw_colors
    ]

    query_tracker.assert_statements(2, "select", sql_snapshot)


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
async def test_empty_query_hook(fruits_query: str, any_query: AnyQueryExecutor, raw_fruits: RawRecordData) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ emptyQueryHook }} }}"))

    assert not result.errors
    assert result.data
    assert len(result.data[fruits_query]) == len(raw_fruits)
    assert result.data[fruits_query] == [{"emptyQueryHook": "success"} for _ in range(len(raw_fruits))]


@pytest.mark.parametrize("query", ["filteredFruits", "filteredFruitsPaginated"])
@pytest.mark.snapshot
async def test_custom_query_hook_where(
    query: str, any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    result = await maybe_async(any_query(f"{{ {query} {{ name }} }}"))

    assert not result.errors
    assert result.data
    assert len(result.data[query]) == 1
    assert result.data[query] == [{"name": "Apple"}]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize("query", ["orderedFruits", "orderedFruitsPaginated"])
@pytest.mark.snapshot
async def test_custom_query_hook_order_by(
    query: str,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    raw_fruits: RawRecordData,
) -> None:
    result = await maybe_async(any_query(f"{{ {query} {{ name }} }}"))

    assert not result.errors
    assert result.data
    assert len(result.data[query]) == len(raw_fruits)
    assert result.data[query] == sorted(
        [{"name": fruit["name"]} for fruit in raw_fruits], key=lambda fruit: fruit["name"]
    )

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_query_hook_on_type(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    result = await maybe_async(any_query("{ colors { fruits { name } } }"))

    assert not result.errors
    assert result.data
    assert len(result.data["colors"]) == 1
    assert result.data["colors"] == [{"fruits": [{"name": "Apple"}]}]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
