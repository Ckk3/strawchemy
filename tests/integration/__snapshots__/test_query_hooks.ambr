# serializer version: 1
# name: test_always_load_column_hook[session-tracked-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_custom_query_hook[session-tracked-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruits-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruits-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruits-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruitsPaginated-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM
    (SELECT fruit.name AS name,
            fruit.id AS id
     FROM fruit AS fruit
     WHERE fruit.name = :name_1
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  WHERE fruit.name = :name_2
  LIMIT :param_3
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruitsPaginated-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM
    (SELECT fruit.name AS name,
            fruit.id AS id
     FROM fruit AS fruit
     WHERE fruit.name = :name_1
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  WHERE fruit.name = :name_2
  LIMIT :param_3
  '''
# ---
# name: test_custom_query_hook[session-tracked-userFruitsPaginated-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM
    (SELECT fruit.name AS name,
            fruit.id AS id
     FROM fruit AS fruit
     WHERE fruit.name = :name_1
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  WHERE fruit.name = :name_2
  LIMIT :param_3
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
